### PR TITLE
RavenDB-14515 - added load revisions lazily implementation

### PR DESCRIPTION
**Pull request by omerlewitz**

### CODE REVIEW

File: src/Raven.Client/Documents/Session/IRevisionsSessionOperationsAsync.cs [hunk 1]
```diff
@@ -64,6 +71,17 @@ public interface IRevisionsSessionOperationsAsync
         /// <summary>
         /// Returns the number of revisions for specified document.
         /// </summary>
-        Task<long> GetCountForAsync(string id, CancellationToken token = default);
+        Task<long> GetCountForAsync(string id, CancellationToken token = default);//
```
Comment (ayende) [inline src/Raven.Client/Documents/Session/IRevisionsSessionOperationsAsync.cs]:
```suggestion
        Task<long> GetCountForAsync(string id, CancellationToken token = default);
```

**Review by ayende — COMMENTED** [review]
Looks good, need to also create sync API to wrap it.

**Review by ayende — APPROVED** [review]

File: src/Raven.Client/Documents/Commands/GetRevisionsCommand.cs [hunk 2]
```diff
@@ -45,18 +44,32 @@ public GetRevisionsCommand(string[] changeVectors, bool metadataOnly = false)
...
+                .Append("/databases/")
                 .Append(node.Database)
                 .Append("/revisions?");
+            GetRequestQueryString(pathBuilder);
+
+            url = pathBuilder.ToString();
+            return request;
+        }
+
+        public string GetRequestQueryString()
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Commands/GetRevisionsCommand.cs]: I think we can switch it to `private` ?

File: src/Raven.Client/Documents/Session/DocumentSessionRevisionsAsync.cs [hunk 3]
```diff
@@ -10,19 +10,25 @@
...
 namespace Raven.Client.Documents.Session
 {
     /// <summary>
     /// Implements Unit of Work for accessing the RavenDB server
     /// </summary>
-    public class DocumentSessionRevisionsAsync : DocumentSessionRevisionsBase, IRevisionsSessionOperationsAsync
+    public class DocumentSessionRevisionsAsync : DocumentSessionRevisionsBase, IRevisionsSessionOperationsAsync, ILazyRevisionsOperationsAsync
     {
-        public DocumentSessionRevisionsAsync(InMemoryDocumentSessionOperations session) : base(session)
+        private new AsyncDocumentSession Session { get; }
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/DocumentSessionRevisionsAsync.cs]: Avoid `new` if possible. Better to cast it inline like you did in sync session.

File: src/Raven.Client/Documents/Session/DocumentSessionRevisionsAsync.cs [hunk 4]
```diff
@@ -93,5 +100,43 @@ public async Task<long> GetCountForAsync(string id, CancellationToken token = de
                 return command.Result;
             }
         }
+
+        Lazy<Task<List<T>>> ILazyRevisionsOperationsAsync.GetForAsync<T>(string id, int start = 0, int pageSize = 25, CancellationToken token = default)
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/DocumentSessionRevisionsAsync.cs]: Can you fix the warnings here?

File: src/Raven.Client/Documents/Session/IRevisionsSessionOperations.cs [hunk 5]
```diff
@@ -74,5 +74,22 @@ public interface IRevisionsSessionOperations
...
+        /// Access the lazy revisions operations 
+        /// </summary>
+        ILazyRevisionsOperations Lazily { get;}
+    }
+    /// <summary>
+    /// Revisions advanced synchronous Lazy session operations
+    /// </summary>
+    public interface ILazyRevisionsOperations
+    {
+        Lazy<List<T>> GetFor<T>(string id, int start = 0, int pageSize = 25);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/IRevisionsSessionOperations.cs:91]: can you copy the comments IRevisionsSessionOperations here? Remember to adjust the text there if needed.

File: src/Raven.Client/Documents/Session/IRevisionsSessionOperationsAsync.cs [hunk 6]
```diff
@@ -65,5 +72,16 @@ public interface IRevisionsSessionOperationsAsync
         /// Returns the number of revisions for specified document.
         /// </summary>
         Task<long> GetCountForAsync(string id, CancellationToken token = default);
+
+    }
+
+    public interface ILazyRevisionsOperationsAsync
+    {
+        Lazy<Task<List<T>>> GetForAsync<T>(string id, int start = 0, int pageSize = 25, CancellationToken token = default);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/IRevisionsSessionOperationsAsync.cs:83]: Can you copy the comments from 'IRevisionsSessionOperationsAsync' ?

**Review by ppekrol — COMMENTED** [review]
Looks good.

File: src/Raven.Client/Documents/Session/IRevisionsSessionOperationsAsync.cs (hunk 1 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/IRevisionsSessionOperationsAsync.cs]: done

File: src/Raven.Client/Documents/Commands/GetRevisionsCommand.cs (hunk 2 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Commands/GetRevisionsCommand.cs]: done

File: src/Raven.Client/Documents/Session/IRevisionsSessionOperations.cs (hunk 5 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/IRevisionsSessionOperations.cs:91]: done

File: src/Raven.Client/Documents/Session/IRevisionsSessionOperationsAsync.cs (hunk 6 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/IRevisionsSessionOperationsAsync.cs:83]: done

File: src/Raven.Client/Documents/Session/DocumentSessionRevisionsAsync.cs (hunk 4 above)
Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/DocumentSessionRevisionsAsync.cs]: done

File: src/Raven.Client/Documents/Session/IRevisionsSessionOperationsAsync.cs (hunk 6 above)
Comment (ppekrol) [reply src/Raven.Client/Documents/Session/IRevisionsSessionOperationsAsync.cs:83]: Ping

Comment (omerlewitz) [reply src/Raven.Client/Documents/Session/IRevisionsSessionOperationsAsync.cs:83]: done

**Pull request merged**